src/test/java/picard/sam/RevertSamTest.java

```diff
@@ -592,4 +604,135 @@ public void testHardClipRoundTrip() throws Exception {
+                {NA12878_MEDIUM, null, OUTPUT_BY_READ_GROUP, OutputMapOptions.CLOUD_OUTPUT_MAP },
+                // Output by read group using the local output map, write output bams in the cloud
+                {NA12878_MEDIUM, null, OUTPUT_BY_READ_GROUP, OutputMapOptions.LOCAL_OUTPUT_MAP },
+                // Cram input, create a cram for each read group
+                {NA12878_MEDIUM_CRAM, null, OUTPUT_BY_READ_GROUP, OutputMapOptions.CLOUD_OUTPUT_MAP },
+                // Cram input, single output cram
+                {NA12878_MEDIUM_CRAM, tempCramOutputPath, !OUTPUT_BY_READ_GROUP, OutputMapOptions.NO_OUTPUT_MAP }};
+    }
+
+    /**
+     *
+     * Write the output map for Revert Sam based on the read groups present in the header file of the given SAM file.
+     *
+     * @param sam        The read names in the header of this sam file will be used as rows
+     * @param outputBase The directory where the per read group output files will live
+     * @return The PicardHtsPath to the tsv with each row a pair (read name, output), where output points to a temp file
+o     * created in the directory specified in outputBase.
+     */
+    private PicardHtsPath getTestReadGroupMapFile(final PicardHtsPath sam, final String outputBase) {
+        final String samExtension = sam.isBam() ? ".bam" : ".cram";
+        final PicardHtsPath result = PicardBucketUtils.getTempFilePath(DEFAULT_CLOUD_TEST_OUTPUT_DIR.getURIString(), ".txt");
+        final List<String> readGroups = getReadGroups(sam.toPath()).stream().map(rg -> rg.getId()).collect(Collectors.toList());
+        final List<String> readGroupOutput = new ArrayList<>();
+
+        for (String readGroup : readGroups){
+            final PicardHtsPath readGroupOutputPath = PicardBucketUtils.getTempFilePath(outputBase, readGroup, samExtension);
+            readGroupOutput.add(readGroupOutputPath.getURIString());
+        }
+
+        RevertSam.writeOutputMapToFile(result.toPath(), readGroups, readGroupOutput);
+        return result;
+    }
+
+    private List<SAMReadGroupRecord> getReadGroups(final Path sam){
+        final List<SAMReadGroupRecord> readGroupsInInput = SamReaderFactory.makeDefault().open(sam).getFileHeader().getReadGroups();
+        return readGroupsInInput;
+    }
+
+    /***
+     *
+     * @param inputSAM The path to the input SAM file
+     * @param outputPath The path to the single output SAM file, when output by read group is turned off. May be null.
+     * @param outputByReadGroup Whether to output by read group
+     * @param outputMapOptions Dictates whether to dynamically create an output map and where (local or cloud)
+     */
+    @Test(dataProvider = "cloudTestData", groups = "cloud")
+    public void testCloud(final PicardHtsPath inputSAM, final PicardHtsPath outputPath, final boolean outputByReadGroup,
+                          final OutputMapOptions outputMapOptions) {
```

Comment:
It took me a while to figure out why all of these args are needed. It's very possible that I'm missing some nuance, but it seems like it all of this (test and provider) code would be a lot simpler if there were just 3 values in the data provider: the input path, the output path (which will either be a file or a directory), and a `writeByReadGroup` boolean (which tells you whether the output path needs to be a map (OUTPUT_MAP) or a file (OUTPUT)). Everything else in the test method could be derived from those, and then you don't need the enum.

Reply:
It might be, but I think it's a bit dangerous to rely on `Files.isDirectory(OUTPUT)`given how directories are handled in GCS. And even if we did that, I don't think the code would be that much simpler. The reason why this code is hard to parse, I think, is because RevertSam accepts a wide array of input patterns (e.g. OUTPUT may be a file or directory), so I vote that we leave it as is. What do you think?